File: analyzer/src/ast/identifier.rs
use crate::{
    ast::AstParse,
    error::Result,
    lexer::{token::Token, token_type::TokenType},
    parser::Parser,
};

#[derive(Debug, Clone)]
pub struct Identifier {
    pub id: String,
}

impl AsRef<str> for Identifier {
    fn as_ref(&self) -> &str {
        return &self.id;
    }
}

impl Identifier {
    pub fn new(id: String) -> Self {
        Self { id }
    }
}

impl AstParse for Identifier {
    fn parse(parser: &mut Parser) -> Result<Self> {
        let token = parser.next_token_and_expect(TokenType::Identifier)?;
        return Ok(Identifier::new(token.text().to_owned()));
    }
}

impl From<&str> for Identifier {
    fn from(value: &str) -> Self {
        return Self::new(value.to_owned());
    }
}

impl From<&Token> for Identifier {
    fn from(value: &Token) -> Self {
        return Self::new(value.text().to_owned());
    }
}

File: src/Parser/Identifier.cpp
#include "Ast/Identifier.hpp"
#include "Parser/Parser.hpp"

namespace Marble
{

    Identifier::Identifier(const std::string &id, const Span &span) : Ast{span, AstType::Identifier}, m_Id{id}
    {
    }

    Identifier::Identifier(std::string &&id, const Span &span) : Ast{span, AstType::Identifier}, m_Id{std::move(id)}
    {
    }

    Identifier::Identifier(const Token &token) : Ast{token.Span(), AstType::Identifier}, m_Id{token.Text()}
    {
    }

    Identifier::Identifier(Token &&token) : Ast{std::move(token.Span()), AstType::Identifier}, m_Id{std::move(token.Text())}
    {
    }

    Identifier::Identifier(const std::string &id, Span &&span) : Ast{std::move(span), AstType::Identifier}, m_Id{id}
    {
    }

    Identifier::Identifier(std::string &&id, Span &&span) : Ast{std::move(span), AstType::Identifier}, m_Id{std::move(id)}
    {
    }

    Identifier::Identifier(const Identifier &obj) : Ast{obj.m_Span, AstType::Identifier}, m_Id{obj.m_Id}
    {
    }

    Identifier::Identifier(Identifier &&obj) : Ast{std::move(obj.m_Span), AstType::Identifier}, m_Id{std::move(obj.m_Id)}
    {
    }

    Identifier &Identifier::operator=(const Identifier &obj)
    {
        if (this == &obj)
        {
            return *this;
        }
        m_Span = obj.m_Span;
        m_Type = obj.m_Type;
        m_Id = obj.m_Id;
        return *this;
    }

    Box<Identifier> Identifier::Parse(Parser &parser)
    {
        const Token &token = parser.NextTokenAndExpect(TokenType::Identifier);
        return MakeBox<Identifier>(token);
    }

} // namespace Marble